src/vm/eventpipebuffermanager.cpp

```diff
@@ -0,0 +1,595 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "eventpipeconfiguration.h"
+#include "eventpipebuffer.h"
+#include "eventpipebuffermanager.h"
+
+#ifdef FEATURE_PERFTRACING
+
+EventPipeBufferManager::EventPipeBufferManager()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    m_pPerThreadBufferList = new SList<SListElem<EventPipeBufferList*>>();
+    m_sizeOfAllBuffers = 0;
+    m_lock.Init(LOCK_TYPE_DEFAULT);
+
+#ifdef _DEBUG
+    m_numBuffersAllocated = 0;
+    m_numBuffersStolen = 0;
+#endif // _DEBUG
+}
+
+EventPipeBuffer* EventPipeBufferManager::AllocateBufferForThread(Thread *pThread, unsigned int requestSize)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        PRECONDITION(pThread != NULL);
+        PRECONDITION(requestSize > 0);
+    }
+    CONTRACTL_END;
+
+    // Allocating a buffer requires us to take the lock.
+    SpinLockHolder _slh(&m_lock);
```

Comment:
There is a lot of stuff under the spinlock, including memory allocations. It would be great to reduce the amount of stuff done under the spinlock to the minimum necessary. It seems that the code can be relatively easily refactored to achieve that goal. Unless I am mistaken, only the m_pPerThreadBufferList->InsertTail and a portion of the code to steal a buffer from another thread needs to be synchronized.